### PR TITLE
BUG: linalg: fix cython import order

### LINKED ISSUE #18909: BUG: Cython3 linalg import order bug
### Describe your issue.

Since Cython 3, I am unable to cythonize if the cimports are in a certain reverse order. Only happens on Linux, not on MacOS. The paths for imports are wrong in the traceback, so this might be a consequence of [this change](https://cython.readthedocs.io/en/latest/src/userguide/migrating_to_cy30.html#namespace-packages).

```
# This doesn't work...
cimport scipy.linalg.cython_lapack
cimport scipy.linalg
```
```
# This works...
cimport scipy.linalg
cimport scipy.linalg.cython_lapack
```

### Reproducing Code Example

```python
# This is in an environment with scipy and cython 3 installed
# In test.pyx, have the following lines:
cimport scipy.linalg.cython_lapack
cimport scipy.linalg

# Run the following
cython test.pyx
```

### Error message

```shell
/home/user/anaconda3/envs/scipy-cy/lib/python3.11/site-packages/Cython/Compiler/Main.py:381: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /home/user/git15/scipy/test.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
from .linalg cimport cython_blas, cython_lapack
^
------------------------------------------------------------

/home/user/anaconda3/envs/scipy-cy/lib/python3.11/site-packages/scipy/linalg.pxd:1:0: 'scipy/linalg/linalg.pxd' not found

Error compiling Cython file:
------------------------------------------------------------
...
from .linalg cimport cython_blas, cython_lapack
^
------------------------------------------------------------

/home/user/anaconda3/envs/scipy-cy/lib/python3.11/site-packages/scipy/linalg.pxd:1:0: 'scipy/linalg/linalg/cython_blas.pxd' not found

Error compiling Cython file:
------------------------------------------------------------
...
from .linalg cimport cython_blas, cython_lapack
^
------------------------------------------------------------

/home/user/anaconda3/envs/scipy-cy/lib/python3.11/site-packages/scipy/linalg.pxd:1:0: 'scipy/linalg/linalg/cython_lapack.pxd' not found
```

### SciPy/NumPy/Python version and system information

```shell
>>> import sys,scipy,numpy,cython; print(scipy.__version__, numpy.__version__, cython.__version__, sys.version_info); scipy.show_config()
1.10.1 1.25.0 3.0.0 sys.version_info(major=3, minor=11, micro=4, releaselevel='final', serial=0)
lapack_armpl_info:
  NOT AVAILABLE
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/user/anaconda3/envs/scipy-cy/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/user/anaconda3/envs/scipy-cy/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/user/anaconda3/envs/scipy-cy/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/user/anaconda3/envs/scipy-cy/include']
blas_armpl_info:
  NOT AVAILABLE
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/user/anaconda3/envs/scipy-cy/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/user/anaconda3/envs/scipy-cy/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/user/anaconda3/envs/scipy-cy/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/user/anaconda3/envs/scipy-cy/include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_KNL,AVX512_KNM,AVX512_SKX,AVX512_CLX,AVX512_CNL,AVX512_ICL
```

### PR DESCRIPTION
**Pull request by lucascolley**
Closes gh-18909, cc @rgommers @jeremyleung521 @mathstuf

### CODE REVIEW

File: scipy/meson.build [hunk 1]
```diff
@@ -379,7 +378,6 @@ py3.install_sources(
 #_cython_tree = declare_dependency(sources: [
 _cython_tree = [
   fs.copyfile('__init__.py'),
-  fs.copyfile('linalg.pxd'),
   fs.copyfile('optimize.pxd'),
   fs.copyfile('special.pxd'),
```
Comment (lucascolley) [inline scipy/meson.build:382]: I suppose we can clean up optimize and special as well?

File: scipy/linalg/tests/_cython_examples/extending.pyx [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #cython: boundscheck=False
 #cython: wraparound=False
 
+cimport scipy.linalg
```
Comment (tylerjereddy) [inline scipy/linalg/tests/_cython_examples/extending.pyx:6]: FWIW, both this and the original reproducer already pass for me locally on Linux with SciPy `main`:

```diff
--- a/scipy/linalg/tests/_cython_examples/extending.pyx
+++ b/scipy/linalg/tests/_cython_examples/extending.pyx
@@ -3,6 +3,8 @@
 #cython: boundscheck=False
 #cython: wraparound=False
 
+cimport scipy.linalg.cython_lapack
+cimport scipy.linalg
 from scipy.linalg.cython_blas cimport cdotu
 from scipy.linalg.cython_lapack cimport dgtsv
```

I don't know if we'll have much luck reproducing in the testsuite, since the reports in the issue seem a bit flaky perhaps.

File: scipy/meson.build (hunk 1 above)
Comment (rgommers) [reply scipy/meson.build:382]: Yes please, treating them all the same is better.

File: scipy/linalg/meson.build [hunk 3]
```diff
@@ -31,19 +32,19 @@ cython_blas_pxd = cython_linalg[2]
 linalg_init_cython_gen = generator(cython,
   arguments : cython_args,
   output : '@BASENAME@.c',
-  depends : [_cython_tree, __init__py])
+  depends : [_cython_tree, __init__py, __init__pxd])
```
Comment (rgommers) [inline scipy/linalg/meson.build:35]: This would introduce a small bug, since other submodules (e.g., `cluster`) also depend on this header. Can you revert this change and the two `depends:` entries further down, and instead use:
```meson
_cython_tree += __init__pxd
```
That will keep the `_cython_tree` content unchanged from what it is now.

**Review by rgommers — COMMENTED** [review]
Thanks @lucascolley for cleaning this up.

**Comment by lucascolley** [comment]
Updated for the other submodules too. But reading gh-18810, is any copying of files supposed to be necessary at all any more? It sounded like it was only introduced due to a bug which has been fixed?

**Comment by rgommers** [comment]
It's still needed. gh-18810 was a specific bug, but in general Cython is not able to handle building from generated sources without either copying over all relevant `.pxd` and `__init__.py` files or by changing the `cython` invocations to add `--module-name`.

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks Lucas! And thanks for testing Tyler.
